sound/soc/sof/intel/hda-dsp.c

```diff
@@ -887,6 +887,67 @@ int hda_dsp_suspend(struct snd_sof_dev *sdev, u32 target_state)
 	return snd_sof_dsp_set_power_state(sdev, &target_dsp_state);
 }
 
+static unsigned int hda_dsp_check_for_dma_streams(struct snd_sof_dev *sdev)
+{
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	struct hdac_stream *s;
+	int active_streams = 0;
+	int sd_offset;
+	u32 val;
+
+	list_for_each_entry(s, &bus->stream_list, list) {
+		sd_offset = SOF_STREAM_SD_OFFSET(s);
+		val = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR,
+				       sd_offset);
+		if (val & SOF_HDA_SD_CTL_DMA_START)
+			active_streams |= BIT(s->index);
+	}
+
+	return active_streams;
+}
+
+static int hda_dsp_s5_quirk(struct snd_sof_dev *sdev)
+{
+	int ret;
+
+	usleep_range(500, 1000);
+
+	ret = hda_dsp_ctrl_link_reset(sdev, false);
+	if (ret < 0)
+		return ret;
+
+	usleep_range(500, 1000);
+
+	ret = hda_dsp_ctrl_link_reset(sdev, true);
+	if (ret < 0)
+		return ret;
```

Comment:
should there be a sleep here to make sure all the hardware is in the right state before rebooting?

Reply:
hda_dsp_ctrl_link_reset() already has a loop to wait until state has changed, so we should be able to trust this and have no test data pointing this to not to be the case. The sleeps between reset out, reset in, I took from other place in the code where we already have such delays (and have proven reliable, like the caps query).

Reply:
ok

Reply:
@kv2019i From the HDA spec, ```Note that the CORB/RIRB RUN bits and all Stream RUN bits must be verified
cleared to 0 before CRST# is written to 0 (asserted) in order to assure a clean restart.``` I think if there are active streams, it should be better to stop the stream firstly.